lnwire/extra_bytes.go

```diff
@@ -50,18 +50,45 @@ func (e *ExtraOpaqueData) Decode(r io.Reader) error {
 // PackRecords attempts to encode the set of tlv records into the target
 // ExtraOpaqueData instance. The records will be encoded as a raw TLV stream
 // and stored within the backing slice pointer.
-func (e *ExtraOpaqueData) PackRecords(recordProducers ...tlv.RecordProducer) error {
-	// First, assemble all the records passed in in series.
-	records := make([]tlv.Record, 0, len(recordProducers))
-	for _, producer := range recordProducers {
-		records = append(records, producer.Record())
+//
+// NOTE: This method takes account of any existing TLV records in the byte
+// slice, and ensures that the resulting set of records maintain a canonical
+// order, with TLV type values sorted in ascending order.
+func (e *ExtraOpaqueData) PackRecords(newRecords ...tlv.RecordProducer) error {
```

Comment:
I think this needs to also take into account the case where a new record has the same same type as an existing record. Otherwise this cannot be called twice which we defs will want to do sometimes. For example, if we decode a message - we will set the ExtraOpaqueData of that message and then when we go to re-encode, this will  duplicate the records instead of overwriting the existing ones. 

we do this a couple of times throughout the code base:

```
err := EncodeMessageExtraData(&q.ExtraData, recordProducers...)
```

where `q` is the message. As is, afaict, this will only work properly if we call `Encode` on a message _once_. If we ever go to encode it twice, it will error with "non canonical" due to the duplicate tlv entries

Reply:
Thanks for taking a look.

> As is, afaict, this will only work properly if we call Encode on a message once. If we ever go to encode it twice, it will error with "non canonical" due to the duplicate tlv entries

I've also seen instances in code where calling `Encode` has the side effect of modifying some field on the struct instance (an `ExtraData` field or similar). I think that sort of side effect is a bad idea.